crates/ironrdp-rdpsnd-native/src/cpal.rs

```diff
@@ -0,0 +1,244 @@
+        let mut filled = 0;
+
+        while filled < data.len() {
+            if self.last.is_none() {
+                match self.receiver.recv_timeout(Duration::from_millis(4000)) {
+                    Ok(rx) => {
+                        debug!("{}", rx.len());
+                        self.last = Some(rx);
+                    }
+                    Err(err) => {
+                        error!(?err);
+                    }
+                }
+            }
+
+            let Some(ref last) = self.last else {
+                info!("playback rx underrun");
+                return;
+            };
+
+            #[allow(clippy::arithmetic_side_effects)]
+            while self.idx < last.len() && filled < data.len() {
+                data[filled] = last[self.idx];
+                assert!(filled < usize::MAX);
+                assert!(self.idx < usize::MAX);
+                filled += 1;
+                self.idx += 1;
+            }
+
+            // If all elements from last have been consumed, clear `self.last`
+            if self.idx >= last.len() {
+                self.last = None;
+                self.idx = 0;
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{sync::mpsc, thread, time::Duration};
+
+    use cpal::traits::StreamTrait;
+    use ironrdp_rdpsnd::pdu::WaveFormat;
+
+    use super::*;
+
+    fn setup_logging() -> anyhow::Result<()> {
+        use tracing::metadata::LevelFilter;
+        use tracing_subscriber::prelude::*;
+        use tracing_subscriber::EnvFilter;
+
+        let fmt_layer = tracing_subscriber::fmt::layer().compact();
+
+        let env_filter = EnvFilter::builder()
+            .with_default_directive(LevelFilter::WARN.into())
+            .with_env_var("IRONRDP_LOG")
+            .from_env_lossy();
+
+        tracing_subscriber::registry()
+            .with(fmt_layer)
+            .with(env_filter)
+            .try_init()
+            .context("failed to set tracing global subscriber")?;
+
+        Ok(())
+    }
+
+    #[test]
+    fn it_works() {
+        setup_logging().unwrap();
+
+        let rx_format = AudioFormat {
+            format: WaveFormat::PCM,
+            n_channels: 2,
+            n_samples_per_sec: 22050,
+            n_avg_bytes_per_sec: 88200,
+            n_block_align: 4,
+            bits_per_sample: 16,
+            data: None,
+        };
+        let (tx, rx) = mpsc::channel();
+
+        let stream = make_stream(&rx_format, rx).unwrap();
+
+        let _producer = thread::spawn(move || {
+            let data_chunks = vec![vec![1u8, 2, 3], vec![4, 5, 6], vec![7, 8, 9]];
+            for chunk in data_chunks {
+                tx.send(chunk).expect("Failed to send data chunk");
+                debug!("sent a chunk");
+                thread::sleep(Duration::from_secs(1)); // Simulating work
+            }
+        });
+
+        stream.play().unwrap();
+        std::thread::sleep(Duration::from_millis(4000));
```

Comment:
issue: It’s not okay for the test to block for 4 seconds.

Reply:
moved as example